Issue#15 grow ethereum hackathon : Update To Latest Seed-Rs

levalleux-ludo opened the pull request:
Hi there,
I've upgraded the project to seed 0.4.0.
Everything looks working as the previous version.

David-OConnor committed:
Update README.md

David-OConnor committed:
Merge branch 'master' of https://github.com/David-OConnor/seed-quickstart

David-OConnor committed:
Updated for 0.2.0

David-OConnor committed:
Updated for window events

David-OConnor committed:
Changes in readme to reflect improved build script

David-OConnor committed:
Updated for v0.2.4 release

David-OConnor committed:
Updated for v0.2.5 release

David-OConnor committed:
Updated for v0.2.6 patch

David-OConnor committed:
Updated version

David-OConnor committed:
Updated for 0.2.9, including a breaking change to Update

David-OConnor committed:
removed clone from model per new Seed version

David-OConnor committed:
Update for 0.3.0

David-OConnor committed:
Fixed problem with build script ps1

frozar committed:
[README] Fix a typo.

frozar committed:
[SCRIPT] Fix the build.sh script.

* Add the execution write over build.sh script.
* Apply the command dos2unix over build.sh script to be able to launch it.

frozar committed:
[SCRIPT] Check the python interpretor version in serve.py

If the python interpretor is to old, it doesn't have yet the context manager
protocol support:
https://stackoverflow.com/questions/48055880/attributeerror-exit-with-socketserver-on-python-3-4-3#answer-48056064

An alternative syntax is proposed to support python interpretor older than 3.6

frozar committed:
[SCRIPT] serve.py: display the address of the URL used by the server (copy/paste).

David-OConnor committed:
Updated for v0.3.1

David-OConnor committed:
Updated for v0.3.2

David-OConnor committed:
Updated to be IAW new build process

David-OConnor committed:
Fixed minor errors in makefile and readme

David-OConnor committed:
Updated comments on makefile

David-OConnor committed:
Switched dev server from https to microserver, for spa routing support

David-OConnor committed:
Added testing commands to Makefile

David-OConnor committed:
Tweaked makefile

David-OConnor committed:
Updated for v0.3.4 release

David-OConnor committed:
Updated for v0.3.5

Commit:
initial commit

Commit:
get market campaigns

Commit:
sum campaigns

Commit:
some basic functionality

Commit:
bit more elegant view_channel_table

Commit:
remove log

Commit:
add missing src/bignum

Commit:
add Paid column

Commit:
add status

Commit:
paid ratio

Commit:
styles

Commit:
is_loading implemented

Commit:
fix loading

Commit:
fix favicon and relative paths

Commit:
fix favicon again

Commit:
favicon, seriously

Commit:
sort channels

Commit:
estimate impressions

Commit:
formatted impression count

Commit:
tied some styling

Commit:
model changes

Commit:
sort

Commit:
format dates

Commit:
update every 30 esconds

Commit:
proper calculation of impressions

Commit:
add total paid

Commit:
copy-paste types from domain cause it uses Futures

Commit:
Show channel URL

Commit:
cargo fmt

Commit:
estimate last updated

Commit:
cargo fmt

Commit:
include exact time

Commit:
README updated

Commit:
a few more stats

Commit:
get balance from etherscan

Commit:
refactored: ActionLoad

Commit:
Refresh msg

Commit:
rename some fns

Commit:
update model to have Channel details

Commit:
routing implemented

Commit:
cards

Commit:
Cards styling

Commit:
remove unused struct

Commit:
add Created

Commit:
sort by created

Commit:
previews

Commit:
paint non-recent items (rather than displaying last checked), less verbose time

elpiel committed:
Cargo.toml - add domain from `adex-validator-stack-rust`

elpiel committed:
README - add `wasm-pack` step

Commit:
try using the domain crate

Commit:
.gitignore: do not ignore lock file

Commit:
Cargo.toml: oops, forgot that

Commit:
preview is now clickable

elpiel committed:
Merge branch 'rust' into rust-domain

elpiel committed:
Use domain crate:
- Cargo.toml - add `domain` crate with local path
- lib - fix code

Commit:
Cargo.toml: use git

Commit:
add CPM

Commit:
index.html: cache break

Commit:
cargo fmt

Commit:
fix ad unit counting

Commit:
ad unit counting improved

Commit:
add text shim, therefore fixing Edge

Commit:
Cargo.lock added

Commit:
removed ad_unit

Commit:
impl Default for the enums

Commit:
compare last updated time properly

Commit:
view functions should be pure

Commit:
refactor: channel_dai is first an iterator, and we then turn it into a vec

Commit:
use lazysort

Commit:
cargo fmt

Commit:
SortSelected: From<String>

Commit:
calculate total publishers

Commit:
cargo fmt

Commit:
move stuff around

Commit:
cleanups

Commit:
cleanup comments

Commit:
Merge pull request #11 from AdExNetwork/rust

Rust re-write

Commit:
move the ad unit chain to unique_units

Commit:
Merge branch 'master' of github.com:AdExNetwork/adex-explorer

Commit:
stats about ad units by size; fixes #14

Commit:
simplified structures in ad_unit_stats_table

Commit:
code style changes: fragments that render tables should return, well, a table

Commit:
update deploy instructions

Commit:
isolate a separate helper for converting URLs to http

Commit:
support previewing videos

Commit:
video preview: loop it

Commit:
add 24hour volume

Commit:
WIP volume graph

Commit:
update domain, compiles now

Commit:
graph sorta works

Commit:
chart styling

Commit:
more styling, although it needs improvemenets

Commit:
final fixes in width

Commit:
Merge pull request #17 from AdExNetwork/24h-volume

volume graph

Commit:
Reordered stats, dropped a need of type inference

Commit:
Add max impressions count

Commit:
hide max impressions, does not seem too useful for now

Commit:
src/lib: remove last element from points

Commit:
refactor volume_chart

Commit:
more refactor

Commit:
show Monthly impressions with a chart

Commit:
cargo fmt

Commit:
Increase refresh_ms

Commit:
loadables

Commit:
refactor loadable stuff

Commit:
more refactors; fixes #19

Commit:
end to end charts

Commit:
video preview: add muted

Commit:
increase REFRESH_MS again

Commit:
more elagant volume_chart

Commit:
Do not show channels by default

Commit:
router fixed, /channels path that shows channels

Commit:
removed total impressions estimation

levalleux-ludo committed:
Upgrade to seed 0.4.0 - compilation OK, Running but a few things to fix

Running quite OK. No deal with errors + view is called twice so it takes long to display and its redondant

Running OK - dealing basically with errors

levalleux-ludo committed:
Create an init() function to force initial route according to the URL

(necessary since seed 0.4.0 - PR #189 'scalable application support'')
https://github.com/David-OConnor/seed/pull/189#issuecomment-517747635

levalleux-ludo committed:
Upgrade other dependencies

levalleux-ludo committed:
clean code from logs and useless comments

Cleaning comments in code

Ivshti reviewed (changes requested):
you can't just `error!`

The whole point of this issue is to handle JSON parse failures elegantly, with the ability to show an error message, rather than just panicking the code

Also, you have a lot of code repetion

make a generic function that will be used to `.map` the `.fetch_json` and emit a new `FetchError` message in the case that it fails

levalleux-ludo commented:
OK, I've gone just straight forward to the upgrade of seed version, taking care of keeping the same behaviour as before, and trying to be minimalist in the code changes.
I will now refactor to get a "better style", and I've a few ideas on how to handle the errors. Could we discuss it on Monday through Discord ?

levalleux-ludo committed:
Merge branch 'master' into issue#15-grow-ethereum-hackathon

levalleux-ludo committed:
refactoring : only one message Msg::DataFetched used by fetch_json, instead of 4

There is now only one message ***Msg::DataFetched*** used by fetch_json (instead of 4).
However we still need to distinguish fetched data by type, regarding the request (the structure to be passed on fetch_json has to be specific to the data received for the JSON to be deserialized in this structure).
For that purpose, the ***Msg::DataFetched*** accepts an ***enum FetchedData*** (4 available values : Balance, MarketChannels, Volume or Impressions).

levalleux-ludo committed:
refactoring : structure FetchDataStruct to contains the FetchObject and a callback to update the model according to the requested information

levalleux-ludo committed:
Improve the fetch error management

we have to deal with the error directly when decoding the FethcObject (inspired from https://github.com/MartinKavik/seed-rs-realworld).
In that way, I’ve created a vector of ’errors’ in Model to push error(s), if any, when decoding.
Then, in ‘view’, the errors are parsed and some HTML elements are created in the page to display the errors.
I’ve used `<details> / <summary>` tags to display errors in a collapsible way.
I’ve also changed `view` a bit : 
 - before : if market_channels fetch fails, nothing is displayed
 - after : every information that has been really fetched is displayed, whereas some others are not, because of the error(s). In case an information can't be computed, 'N/A' is displayed instead of the value.

levalleux-ludo committed:
Refactoring to reduce redundancy + improve Fetch Error management and display errors in page

levalleux-ludo committed:
Improve comments

levalleux-ludo reviewed (commented):
Hi Ivo, 
Following your feedback, here is what I propose :

-	**I have refactored the code in order to avoid redundancy**:
There is now only one message ***Msg::DataFetched*** used by fetch_json (instead of 4).
However we still need to distinguish fetched data by type, regarding the request (the structure to be passed on fetch_json has to be specific to the data received for the JSON to be deserialized in this structure).
For that purpose, the ***Msg::DataFetched*** accepts an ***enum FetchedData*** (4 available values : Balance, MarketChannels, Volume or Impressions).
```rust
enum FetchedData {
        // for requesting balance
        Balance(FetchDataStruct<EtherscanBalResp>),
        // for requesting market_channels
        MarketChannels(FetchDataStruct<Vec<MarketChannel>>),
        // for requesting volume
        Volume(FetchDataStruct<VolumeResp>),
        // for requesting impressions
        Impressions(FetchDataStruct<VolumeResp>),
}
... 
enum Msg {
        Load(ActionLoad),
        Refresh,
        // DataFetched is the unique message to be notified after a fetch request. However the FetchedData value indicates which type of information has been requested
        DataFetched(FetchedData),
        SortSelected(String),
}
```
The enum ***FetchedData*** takes itself a ***struct FetchDataStruct*** that contains the FetchObject returned by fetch_json(), along with some callbacks to define what to do in case of success (callback ‘update_model’) or of error (callback ‘on_error’).
```javascript
struct FetchDataStruct<T> {
        pub fetch_object: fetch::FetchObject<T>,
        // update_model callback is called if the fetch response is OK, in order to update the model according to the fetch data
        pub update_model: fn(&mut Model, Loadable<T>),
        // on_error callback is called if the fetch response in NOK, in order to deal with the error according to the information requested
        pub on_error: fn(&mut Model, Error),
}
```

Doing that, we still can define specific treatments in both cases (fetch OK or KO), but **“decoding” the FetchObject is only coded once** (whereas the real type of FetchObject differs according to the request).

-	**Regarding the improvement of errors management**:
You were proposing to use a message (ex : OnFetchError, as suggested in some seed examples).
But after the upgrade in 0.4.0, the new fetch interface does not require any callback mechanism for the error case, just because the returned FetchObject includes the error, if any.
So **we have to deal with the error directly when decoding the FethcObject** (inspired from https://github.com/MartinKavik/seed-rs-realworld).
In that way, I’ve created a vector of ’errors’ in Model to push error(s), if any, when decoding.
```rust
struct Error {
        title: String,
        descr: String,
}
struct Model {
        ...
        pub errors: Vec<Error>,
}
```
Then, in ‘view’, the errors are parsed and some HTML elements are created in the page to display the errors.
```rust
fn view_errors(model: &Model) -> Node<Msg> {
            if model.errors.len() > 0 {
            div![
                class!["error-messages"],
                    h2!["You have errors !"],
                model.errors.iter().map(|error|
                    div![
                        details![
                            summary![class!["error-summary"], error.title],
                            div![class!["error-details"], error.descr]
                        ]
                    ]
                ),
            ]
        } else {
            empty![]
        }
}
```
I’ve used `<details> / <summary>` tags to display errors in a collapsible way.
I’ve also changed **`view`** a bit : 
 - *before* : if market_channels fetch fails, nothing is displayed
 - *after* : every information that has been really fetched is displayed, whereas some others are not, because of the error(s). In case an information can't be computed, **N/A** is displayed instead of the value.
For that purpose, I’ve added a value `Error` in `Loadable<T>` : 
```rust
enum Loadable<T> {
        Loading,
        Ready(T),
        Error,
}
```
And, for instance, here is what `card` becomes for dealing the error case :
```rust
fn card(label: &str, value: Loadable<String>) -> Node<Msg> {
        div![
            class!["card"],
            match value {
                Loading => div![class!["card-value loading"]],
                Ready(value) => div![class!["card-value"], value],
                Error => {
                    div![class!["card-label"], "N/A"]
                },
            },
            div![class!["card-label"], label],
        ]
}
```

Hope you enjoy my changes.
I stay available to discuss, of course.
